

<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>pplacer &mdash; pplacer 1.1 documentation</title>
    
    <link rel="stylesheet" href="../_static/sphinxdoc.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '1.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <link rel="top" title="pplacer 1.1 documentation" href="../index.html" />
    <link rel="next" title="guppy" href="guppy.html" />
    <link rel="prev" title="pplacer documentation home" href="../index.html" /> 
  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="guppy.html" title="guppy"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="../index.html" title="pplacer documentation home"
             accesskey="P">previous</a> |</li>
        <li><a href="../index.html">pplacer 1.1 documentation</a> &raquo;</li> 
      </ul>
    </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">pplacer</a><ul>
<li><a class="reference internal" href="#options">Options</a></li>
<li><a class="reference internal" href="#details">Details</a><ul>
<li><a class="reference internal" href="#introduction">Introduction</a></li>
<li><a class="reference internal" href="#migrating-from-pplacer-v1-0">Migrating from pplacer v1.0</a></li>
<li><a class="reference internal" href="#pre-masking">Pre-masking</a></li>
<li><a class="reference internal" href="#json-format-specification">JSON format specification</a></li>
<li><a class="reference internal" href="#making-alignments-for-use-with-pplacer">Making alignments for use with pplacer</a></li>
<li><a class="reference internal" href="#making-reference-trees">Making reference trees</a></li>
</ul>
</li>
<li><a class="reference internal" href="#fasttree">FastTree</a><ul>
<li><a class="reference internal" href="#nucleotide-alignments">Nucleotide alignments</a></li>
<li><a class="reference internal" href="#amino-acid-alignments">Amino Acid alignments</a></li>
</ul>
</li>
<li><a class="reference internal" href="#phyml-and-raxml">phyml and RAxML</a><ul>
<li><a class="reference internal" href="#baseball">Baseball</a></li>
<li><a class="reference internal" href="#fig-ranking">Fig ranking</a></li>
</ul>
</li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="../index.html"
                        title="previous chapter">pplacer documentation home</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="guppy.html"
                        title="next chapter">guppy</a></p>
  <h3>This Page</h3>
  <ul class="this-page-menu">
    <li><a href="../_sources/generated_rst/pplacer.txt"
           rel="nofollow">Show Source</a></li>
  </ul>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="pplacer">
<span id="id1"></span><h1>pplacer<a class="headerlink" href="#pplacer" title="Permalink to this headline">¶</a></h1>
<p><cite>pplacer</cite> performs phylogenetic placement.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">pplacer</span> <span class="p">[</span><span class="n">options</span><span class="p">]</span> <span class="p">[</span><span class="n">alignment</span><span class="p">]</span>
</pre></div>
</div>
<div class="section" id="options">
<h2>Options<a class="headerlink" href="#options" title="Permalink to this headline">¶</a></h2>
<table class="docutils option-list" frame="void" rules="none">
<col class="option" />
<col class="description" />
<tbody valign="top">
<tr><td class="option-group">
<kbd><span class="option">-c</span></kbd></td>
<td>Specify the path to the reference package.</td></tr>
<tr><td class="option-group">
<kbd><span class="option">-t</span></kbd></td>
<td>Specify the reference tree filename.</td></tr>
<tr><td class="option-group">
<kbd><span class="option">-r</span></kbd></td>
<td>Specify the reference alignment filename.</td></tr>
<tr><td class="option-group">
<kbd><span class="option">-s</span></kbd></td>
<td>Supply a phyml stats.txt or a RAxML info file giving the model parameters.</td></tr>
<tr><td class="option-group">
<kbd><span class="option">-d</span></kbd></td>
<td>Specify the directory containing the reference information.</td></tr>
<tr><td class="option-group">
<kbd><span class="option">-p</span></kbd></td>
<td>Calculate posterior probabilities.</td></tr>
<tr><td class="option-group">
<kbd><span class="option">-m</span></kbd></td>
<td>Substitution model. Protein: are LG, WAG, or JTT. Nucleotides: GTR.</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--model-freqs</span></kbd></td>
<td>Use model frequencies instead of reference alignment frequencies.</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--gamma-cats</span></kbd></td>
<td>Number of categories for discrete gamma model.</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--gamma-alpha</span></kbd></td>
<td>Specify the shape parameter for a discrete gamma model.</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--ml-tolerance</span></kbd></td>
<td>1st stage branch len optimization tolerance (2nd stage to 1e-5). Default: 0.01.</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--pp-rel-err</span></kbd></td>
<td>Relative error for the posterior probability calculation. Default is 0.01.</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--unif-prior</span></kbd></td>
<td>Use a uniform prior rather than exponential.</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--inform-prior</span></kbd></td>
<td>Use an informative exponential prior based on rooted distance to leaves.</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--prior-lower</span></kbd></td>
<td>Lower bound for the informative prior mean. Default is 0.</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--start-pend</span></kbd></td>
<td>Starting pendant branch length. Default is 0.1.</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--max-pend</span></kbd></td>
<td>Set the maximum ML pendant branch length. Default is 2.</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--fig-cutoff</span></kbd></td>
<td>The cutoff for determining figs. Default is 0; specify 0 to disable.</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--fig-eval-all</span></kbd></td>
<td>Evaluate all likelihoods to ensure that the best location was selected.</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--fig-eval-discrepancy-tree</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>Write out a tree showing the discrepancies between the best complete and observed locations.</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--fig-tree</span></kbd></td>
<td>Write out a tree showing the figs on the tree.</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--max-strikes</span></kbd></td>
<td>Maximum number of strikes for baseball. 0 -&gt; no ball playing. Default is 6.</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--strike-box</span></kbd></td>
<td>Set the size of the strike box in log likelihood units. Default is 3.</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--max-pitches</span></kbd></td>
<td>Set the maximum number of pitches for baseball. Default is 40.</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--fantasy</span></kbd></td>
<td>Desired likelihood cutoff for fantasy baseball mode. 0 -&gt; no fantasy.</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--fantasy-frac</span></kbd></td>
<td>Fraction of fragments to use when running fantasy baseball. Default is 0.1.</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--write-masked</span></kbd></td>
<td>Write alignment masked to the region without gaps in the query.</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--verbosity</span></kbd></td>
<td>Set verbosity level. 0 is silent, and 2 is quite a lot. Default is 1.</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--out-dir</span></kbd></td>
<td>Specify the directory to write place files to.</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--pretend</span></kbd></td>
<td>Only check out the files then report. Do not run the analysis.</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--check-like</span></kbd></td>
<td>Write out the likelihood of the reference tree, calculated two ways.</td></tr>
<tr><td class="option-group">
<kbd><span class="option">-j</span></kbd></td>
<td>The number of child processes to spawn when doing placements. Default is 2.</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--timing</span></kbd></td>
<td>Display timing information after the pplacer run finishes.</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--no-pre-mask</span></kbd></td>
<td>Don&#8217;t pre-mask sequences before placement.</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--write-pre-masked</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>Write out the pre-masked sequences to the specified fasta file and exit.</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--map-mrca</span></kbd></td>
<td>Specify a file to write out MAP sequences for MRCAs and corresponding placements.</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--map-mrca-min</span></kbd></td>
<td>Specify cutoff for inclusion in MAP sequence file. Default is 0.8.</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--map-identity</span></kbd></td>
<td>Add the percent identity of the query sequence to the nearest MAP sequence to each placement.</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--keep-at-most</span></kbd></td>
<td>The maximum number of placements we keep. Default is 7.</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--keep-factor</span></kbd></td>
<td>Throw away anything that has ml_ratio below keep_factor times (best ml_ratio). Default is 0.01.</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--mrca-class</span></kbd></td>
<td>Classify with MRCAs instead of a painted tree.</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--version</span></kbd></td>
<td>Write out the version number and exit.</td></tr>
</tbody>
</table>
</div>
<div class="section" id="details">
<h2>Details<a class="headerlink" href="#details" title="Permalink to this headline">¶</a></h2>
<div class="section" id="introduction">
<h3>Introduction<a class="headerlink" href="#introduction" title="Permalink to this headline">¶</a></h3>
<p>Pplacer places query sequences on a fixed reference phylogenetic tree according to a reference alignment.
In maximum likelihood (ML) mode, pplacer tries to find the attachment location and the pendant branch length which maximize the likelihood of the tree with pendant branch length attached.
In Bayesian posterior probability (PP) mode, pplacer tries to find the edge attachment which maximizes the posterior probability of a fragment placement on an edge conditioned on the reference tree (with branch lengths).</p>
<p>A basic pplacer run looks like:</p>
<div class="highlight-python"><pre>pplacer -c my.refpkg aln.fasta</pre>
</div>
<p>with a reference package.
Running pplacer with a reference package is simple and produces taxonomic annotations which can be used for classification and visualization (see, for example <a class="reference external" href="guppy_classify.html">classify</a> and <a class="reference external" href="guppy_fat.html">fat</a>).
If you wish to override specific parts of the reference package, you can do so with command line options.</p>
<p>A &#8220;reference package&#8221; is simply a collection of files including a reference tree, a reference alignment, and associated taxonomic information.
We have the beginnings of a <a class="reference external" href="http://microbiome.fhcrc.org/apps/refpkg/">reference package database</a> which we hope in the future to be a comprehensive resource for phylogenetic placement.
It&#8217;s just started, and we would love to have your submissions.
For now most users will have to make a reference package using our <a class="reference external" href="http://github.com/fhcrc/taxtastic/">taxtastic</a> package to take advantage of the taxonomic annotation features of pplacer.</p>
<p>The <tt class="docutils literal"><span class="pre">aln.fasta</span></tt> is the alignment file.
It should have the reference sequences which were used to make the reference tree, aligned with the query sequences to be placed on the reference tree.
The reference sequences must have identical names to those used in the reference tree.
It&#8217;s possible to split the alignment into two files (the reference sequences and the query sequences); in that case you need to make sure that those alignments are in the same frame and have the same length.
This splitting was obligatory in v1.0.
The alignment can be in FASTA format (with a <tt class="docutils literal"><span class="pre">.fasta</span></tt> or <tt class="docutils literal"><span class="pre">.fa</span></tt> suffix) or Stockholm format (with a <tt class="docutils literal"><span class="pre">.sto</span></tt> or <tt class="docutils literal"><span class="pre">.sth</span></tt> suffix).</p>
<p>Another way to run pplacer is without a reference package:</p>
<div class="highlight-python"><pre>pplacer -t reference_tree -s statistics_file aln.fasta</pre>
</div>
<p>The <tt class="docutils literal"><span class="pre">statistics_file</span></tt> is a file describing the evolutionary model used to make the reference tree (described in the section on reference tree below).
Running pplacer in this way will disable the taxonomic annotation features of pplacer v1.1.</p>
</div>
<div class="section" id="migrating-from-pplacer-v1-0">
<h3>Migrating from pplacer v1.0<a class="headerlink" href="#migrating-from-pplacer-v1-0" title="Permalink to this headline">¶</a></h3>
<p>There are a couple of differences between the present version and the previous version of pplacer which are worth knowing about.</p>
<ul class="simple">
<li>Rather than having a plain text &#8221;.place&#8221; output file, we now have a <a class="reference external" href="http://www.json.org/">JSON</a>-format file which contains the placements</li>
<li>Reference packages encapsulate reference information, making it easy to do placement and taxonomic annotation</li>
<li>Alignments can now be supplied as a single file, rather than being split into reference and query alignments</li>
<li>Much better: faster, taxonomic integration, etc</li>
<li><tt class="docutils literal"><span class="pre">placeviz</span></tt>, <tt class="docutils literal"><span class="pre">placeutil</span></tt> and <tt class="docutils literal"><span class="pre">mokaphy</span></tt> have been replaced by a single binary called <tt class="docutils literal"><span class="pre">guppy</span></tt></li>
<li><tt class="docutils literal"><span class="pre">rppr</span></tt> binary for preparing reference packages</li>
</ul>
</div>
<div class="section" id="pre-masking">
<h3>Pre-masking<a class="headerlink" href="#pre-masking" title="Permalink to this headline">¶</a></h3>
<p>Columns that are all gap in either the reference alignment or the query alignment do not impact the relative likelihood of placements.
Thus, starting in v1.1alpha08, we don&#8217;t compute them at all.
This speeds things up a bunch, and uses a lot less memory.
We call this pre-masking, and it can be disabled with the <tt class="docutils literal"><span class="pre">--no-pre-mask</span></tt> flag.</p>
</div>
<div class="section" id="json-format-specification">
<h3><a class="reference external" href="http://www.json.org/">JSON</a> format specification<a class="headerlink" href="#json-format-specification" title="Permalink to this headline">¶</a></h3>
<p>The new JSON format is very simple. Each document is a JSON object with a minimum of four keys: <tt class="docutils literal"><span class="pre">tree</span></tt>, <tt class="docutils literal"><span class="pre">fields</span></tt>,
<tt class="docutils literal"><span class="pre">placements</span></tt>, and <tt class="docutils literal"><span class="pre">version</span></tt>. Another key, <tt class="docutils literal"><span class="pre">metadata</span></tt>, is optional. Other keys in the root object are ignored.</p>
<table border="1" class="docutils">
<colgroup>
<col width="18%" />
<col width="82%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Key</th>
<th class="head">Value</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td><tt class="docutils literal"><span class="pre">version</span></tt></td>
<td>The version of the JSON format as an integer.</td>
</tr>
<tr class="row-odd"><td><tt class="docutils literal"><span class="pre">tree</span></tt></td>
<td>The reference tree as a string, in &#8220;edge-numbered Newick&#8221; format.</td>
</tr>
<tr class="row-even"><td><tt class="docutils literal"><span class="pre">placements</span></tt></td>
<td>An array of placements.</td>
</tr>
<tr class="row-odd"><td><tt class="docutils literal"><span class="pre">fields</span></tt></td>
<td>An array of strings corresponding to the data given in the placements array.</td>
</tr>
<tr class="row-even"><td><tt class="docutils literal"><span class="pre">metadata</span></tt></td>
<td>An object containing metadata about the generation of this collection of placements.</td>
</tr>
</tbody>
</table>
<p>An &#8220;edge-numbered Newick&#8221; tree is simply a Newick format tree with integers in
curly braces which provide a well-defined numbering of edges. These edge
numbers are used to specify the edges on which the placements lie.</p>
<p>Currently there are two versions of the placefile format accepted by <tt class="docutils literal"><span class="pre">guppy</span></tt>
and <tt class="docutils literal"><span class="pre">rppr</span></tt>: <tt class="docutils literal"><span class="pre">1</span></tt>, and <tt class="docutils literal"><span class="pre">2</span></tt>, though only version 2 will be generated. There
are only two differences between versions 1 and 2: the format of the Newick
tree in the <tt class="docutils literal"><span class="pre">tree</span></tt> field has changed, and <tt class="docutils literal"><span class="pre">marginal_prob</span></tt> was renamed to
<tt class="docutils literal"><span class="pre">marginal_like</span></tt> in version 2.</p>
<p>Version 1 used a slightly different version of edge-numbered Newick trees for
the <tt class="docutils literal"><span class="pre">tree</span></tt> field, where edge numbers were specified in square brackets
instead of curly braces. Both this kind of Newick tree and version 1 of the
JSON format are now deprecated.</p>
<p>The pplacer suite currently uses the following field names:</p>
<table border="1" class="docutils">
<colgroup>
<col width="19%" />
<col width="81%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Field</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td><tt class="docutils literal"><span class="pre">edge_num</span></tt></td>
<td>The edge number from the provided <tt class="docutils literal"><span class="pre">tree</span></tt> as an integer.</td>
</tr>
<tr class="row-odd"><td><tt class="docutils literal"><span class="pre">likelihood</span></tt></td>
<td>ML log likelihood as a float.</td>
</tr>
<tr class="row-even"><td><tt class="docutils literal"><span class="pre">like_weight_ratio</span></tt></td>
<td>ML likelihood weight ratio as a float.</td>
</tr>
<tr class="row-odd"><td><tt class="docutils literal"><span class="pre">distal_length</span></tt></td>
<td>ML distance from the distal side of the edge as a float.</td>
</tr>
<tr class="row-even"><td><tt class="docutils literal"><span class="pre">pendant_length</span></tt></td>
<td>ML pendant branch length as a float.</td>
</tr>
<tr class="row-odd"><td><tt class="docutils literal"><span class="pre">post_prob</span></tt></td>
<td>The posterior probability of a placement on the edge.</td>
</tr>
<tr class="row-even"><td><tt class="docutils literal"><span class="pre">marginal_like</span></tt></td>
<td>The marginal likelihood of a placement on the edge.</td>
</tr>
<tr class="row-odd"><td><tt class="docutils literal"><span class="pre">classification</span></tt></td>
<td>The <tt class="docutils literal"><span class="pre">tax_id</span></tt> from a reference package as a string.</td>
</tr>
<tr class="row-even"><td><tt class="docutils literal"><span class="pre">map_ratio</span></tt></td>
<td>The percent identity between this sequence and the corresponding MAP sequence.</td>
</tr>
<tr class="row-odd"><td><tt class="docutils literal"><span class="pre">map_overlap</span></tt></td>
<td>The number of overlapping sites between this sequence and the corresponding MAP sequence.</td>
</tr>
</tbody>
</table>
<p>For <tt class="docutils literal"><span class="pre">guppy</span></tt> to be able to load a JSON file, it must have <tt class="docutils literal"><span class="pre">edge_num</span></tt>,
<tt class="docutils literal"><span class="pre">likelihood</span></tt>, <tt class="docutils literal"><span class="pre">like_weight_ratio</span></tt>, <tt class="docutils literal"><span class="pre">distal_length</span></tt>, and
<tt class="docutils literal"><span class="pre">pendant_length</span></tt> fields. All other fields are optional, but if one of
<tt class="docutils literal"><span class="pre">post_prob</span></tt> and <tt class="docutils literal"><span class="pre">marginal_like</span></tt> are specified, both must be specified.</p>
<p>Each entry in the <tt class="docutils literal"><span class="pre">placements</span></tt> array is an object with the following keys:</p>
<table border="1" class="docutils">
<colgroup>
<col width="4%" />
<col width="96%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Key</th>
<th class="head">Value</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td><tt class="docutils literal"><span class="pre">n</span></tt></td>
<td>A string or array of strings corresponding to the name or names of the sequences placed here.</td>
</tr>
<tr class="row-odd"><td><tt class="docutils literal"><span class="pre">p</span></tt></td>
<td>An array of arrays containing placement data in the same order as <tt class="docutils literal"><span class="pre">fields</span></tt>.</td>
</tr>
<tr class="row-even"><td><tt class="docutils literal"><span class="pre">m</span></tt></td>
<td>(optional) A float that represents the mass of this placement. If this key is specified, <tt class="docutils literal"><span class="pre">n</span></tt> must only be or contain a single string.</td>
</tr>
</tbody>
</table>
<p>An example JSON document follows, with the first placement showing uncertainty
in location, and the second showing two reads that had identical placements:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="p">{</span>
  <span class="s">&quot;tree&quot;</span><span class="p">:</span> <span class="s">&quot;((A:2[0],B:9[1]):7[2],C:5[3],D:1[4]):0[5];&quot;</span><span class="p">,</span>
  <span class="s">&quot;placements&quot;</span><span class="p">:</span> <span class="p">[</span>
    <span class="p">{</span><span class="s">&quot;p&quot;</span><span class="p">:</span>
      <span class="p">[[</span><span class="mi">0</span><span class="p">,</span> <span class="o">-</span><span class="mf">1091.576026</span><span class="p">,</span> <span class="mf">0.762438</span><span class="p">,</span> <span class="mf">0.000008</span><span class="p">,</span> <span class="mf">0.019642</span><span class="p">],</span>
        <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="o">-</span><span class="mf">10982.742117</span><span class="p">,</span> <span class="mf">0.237562</span><span class="p">,</span> <span class="mf">0.000008</span><span class="p">,</span> <span class="mf">0.019579</span><span class="p">]</span>
      <span class="p">],</span> <span class="s">&quot;n&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s">&quot;GLKT0ZE01CQ1P1&quot;</span><span class="p">]</span>
    <span class="p">},</span>
    <span class="p">{</span><span class="s">&quot;p&quot;</span><span class="p">:</span> <span class="p">[[</span><span class="mi">2</span><span class="p">,</span> <span class="o">-</span><span class="mf">1061.467623</span><span class="p">,</span> <span class="mf">1.0</span><span class="p">,</span> <span class="mf">0.003555</span><span class="p">,</span> <span class="mf">0.000006</span><span class="p">]],</span>
     <span class="s">&quot;n&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s">&quot;GLKT0ZE01C36CH&quot;</span><span class="p">,</span> <span class="s">&quot;GLKT0ZE01A0IBO&quot;</span><span class="p">]}</span>
  <span class="p">],</span>
  <span class="s">&quot;metadata&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s">&quot;invocation&quot;</span><span class="p">:</span> <span class="s">&quot;guppy to_json&quot;</span><span class="p">},</span>
  <span class="s">&quot;version&quot;</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span>
  <span class="s">&quot;fields&quot;</span><span class="p">:</span> <span class="p">[</span>
    <span class="s">&quot;edge_num&quot;</span><span class="p">,</span> <span class="s">&quot;likelihood&quot;</span><span class="p">,</span> <span class="s">&quot;like_weight_ratio&quot;</span><span class="p">,</span> <span class="s">&quot;distal_length&quot;</span><span class="p">,</span>
    <span class="s">&quot;pendant_length&quot;</span>
  <span class="p">]</span>
<span class="p">}</span>
</pre></div>
</div>
</div>
<div class="section" id="making-alignments-for-use-with-pplacer">
<h3>Making alignments for use with pplacer<a class="headerlink" href="#making-alignments-for-use-with-pplacer" title="Permalink to this headline">¶</a></h3>
<p>There are several options and formats for providing alignments of reference and query sequences.
Examples below illustrate various steps in the sequence alignment process.</p>
<div class="section" id="examples-using-infernal">
<h4>Examples using Infernal<a class="headerlink" href="#examples-using-infernal" title="Permalink to this headline">¶</a></h4>
<p><a class="reference external" href="http://infernal.janelia.org/">Infernal</a> is an excellent package for searching and aligning sequences using RNA secondary structure information.</p>
<div class="section" id="creating-a-reference-alignment">
<h5>Creating a reference alignment<a class="headerlink" href="#creating-a-reference-alignment" title="Permalink to this headline">¶</a></h5>
<p>The first step in any pipeline involving Infernal (assuming you already have an alignment profile but are not working from a reference package) is to create an alignment of reference sequences.
See the Infernal docs for a description of options not mentioned here.
For example:</p>
<div class="highlight-python"><pre>cmalign --hbanded --sub --dna -1 -o refalign.sto profile.cm refseqs.fasta</pre>
</div>
<p>Inputs to this command include an alignment profile (<tt class="docutils literal"><span class="pre">profile.cm</span></tt>) and unaligned reference sequences (<tt class="docutils literal"><span class="pre">refs.fasta</span></tt>).
The output file, identified using the <tt class="docutils literal"><span class="pre">-o</span></tt> option, contains the aligned reference sequences in Stockholm format.
The <tt class="docutils literal"><span class="pre">-1</span></tt> (that&#8217;s a one, not an L) specifies non-interleaved output, one sequence per line.</p>
</div>
<div class="section" id="merging-reference-and-query-sequences">
<h5>Merging reference and query sequences<a class="headerlink" href="#merging-reference-and-query-sequences" title="Permalink to this headline">¶</a></h5>
<p>Query sequences must be aligned with respect to the reference sequences.
This is easily accomplished using two calls to cmalign.
First, align the query sequences just like the reference sequences above:</p>
<div class="highlight-python"><pre>cmalign --hbanded --sub --dna -1 -o qalign.sto profile.cm qseqs.fasta</pre>
</div>
<p>Next, merge the reference and query alignments using the <tt class="docutils literal"><span class="pre">--merge</span></tt> option:</p>
<div class="highlight-python"><pre>cmalign --merge --hbanded --sub --dna -1 -o merged.sto profile.cm refalign.sto qalign.sto</pre>
</div>
<p>Now <tt class="docutils literal"><span class="pre">merged.sto</span></tt> contains a single alignment of both reference and query sequences, and can be used with pplacer as follows after making a reference tree and accompanying statistics file:</p>
<div class="highlight-python"><pre>pplacer -t reference_tree -s statistics_file merged.sto</pre>
</div>
</div>
<div class="section" id="using-a-reference-package">
<h5>Using a reference package<a class="headerlink" href="#using-a-reference-package" title="Permalink to this headline">¶</a></h5>
<p>A closely related example involves alignment with the profile and reference sequences included in a reference package (<tt class="docutils literal"><span class="pre">my.refpkg</span></tt> - note that names may vary in a reference package).
So now we skip creation of the reference alignment.
First, create the query alignment:</p>
<div class="highlight-python"><pre>cmalign --hbanded --sub --dna -1 -o qalign.sto my.refpkg/profile.cm qseqs.fasta</pre>
</div>
<p>...then merge:</p>
<div class="highlight-python"><pre>cmalign --merge --hbanded --sub --dna -1 \
  -o mergedWithRefpkg.sto \
  my.refpkg/profile.cm my.refpkg/refalign.sto qalign.sto</pre>
</div>
<p>Now it is even easier to write the pplacer command:</p>
<div class="highlight-python"><pre>pplacer -c my.refpkg mergedWithRefpkg.sto</pre>
</div>
</div>
</div>
<div class="section" id="examples-using-hmmer">
<h4>Examples using HMMER<a class="headerlink" href="#examples-using-hmmer" title="Permalink to this headline">¶</a></h4>
<p><a class="reference external" href="http://hmmer.janelia.org/">HMMER</a> is another excellent package for searching and aligning sequences by the Eddy group, which can align amino acid and nucleotide sequences.</p>
<p>Assume that we have a reference alignment <tt class="docutils literal"><span class="pre">refseqs.sto</span></tt> in Stockholm format. We first build an HMM:</p>
<div class="highlight-python"><pre>hmmbuild refseqs.hmm refseqs.sto</pre>
</div>
<p>Then we can use it to make a combined alignment with the reference sequences and the reads:</p>
<div class="highlight-python"><pre>hmmalign -o combo.sto --mapali refseqs.sto refseqs.hmm qseqs.fasta</pre>
</div>
<p>Now we can run pplacer:</p>
<div class="highlight-python"><pre>pplacer -t rpoB.tre -s RAxML_info.rpoB combo.sto</pre>
</div>
<p>... or with a reference package:</p>
<div class="highlight-python"><pre>pplacer -c rpoB.refpkg combo.sto</pre>
</div>
</div>
</div>
<div class="section" id="making-reference-trees">
<h3>Making reference trees<a class="headerlink" href="#making-reference-trees" title="Permalink to this headline">¶</a></h3>
</div>
</div>
<div class="section" id="fasttree">
<h2>FastTree<a class="headerlink" href="#fasttree" title="Permalink to this headline">¶</a></h2>
<div class="section" id="nucleotide-alignments">
<h3>Nucleotide alignments<a class="headerlink" href="#nucleotide-alignments" title="Permalink to this headline">¶</a></h3>
<p>FastTree should be used in the following way when making nucleotide reference trees for use with pplacer:</p>
<div class="highlight-python"><pre>FastTree -nt -gtr -log vaginal.log vaginal.fasta &gt; vaginal.tre</pre>
</div>
<p>In particular, do not use the <tt class="docutils literal"><span class="pre">-gamma</span></tt> option, but do use the <tt class="docutils literal"><span class="pre">-gtr</span></tt> option.</p>
</div>
<div class="section" id="amino-acid-alignments">
<h3>Amino Acid alignments<a class="headerlink" href="#amino-acid-alignments" title="Permalink to this headline">¶</a></h3>
<p>FastTree should be used in the following way when making amino acid reference trees for use with pplacer:</p>
<div class="highlight-python"><pre>FastTree -log TIGR00001.log TIGR00001.fasta &gt; TIGR00001.tre</pre>
</div>
<p>Again, <tt class="docutils literal"><span class="pre">-gamma</span></tt> should not be used.</p>
</div>
</div>
<div class="section" id="phyml-and-raxml">
<h2>phyml and RAxML<a class="headerlink" href="#phyml-and-raxml" title="Permalink to this headline">¶</a></h2>
<p><a class="reference external" href="http://www.atgc-montpellier.fr/phyml/&quot;&gt;Phyml&lt;/a&gt;and">PHYML</a> and <a class="reference external" href="http://icwww.epfl.ch/~stamatak/index-Dateien/Page443.htm&quot;&gt;RAxML&lt;/a&gt;">RAxML</a> are two nice packages for making ML trees that are supported for use with pplacer.
Pplacer only knows about the GTR, WAG, LG, and JTT models, so use those to build your trees.
If you are fond of another model and can convince me that I should implement it, I will.</p>
<p>Both of these packages implement gamma rate variation among sites, which accomodates that some regions evolve more quickly than others.
That&#8217;s generally a good thing, but if you have millions of query sequences, you might have to run pplacer with fewer rate parameters to make it faster.</p>
<p>I run RAxML like so, on similar alignments (the &#8220;F&#8221; suffix on PROTGAMMAWAGF means to use the emperical amino acid frequencies):</p>
<div class="highlight-python"><pre>raxmlHPC -m GTRGAMMA -n test -s nucleotides.phy
raxmlHPC -m PROTGAMMAWAGF -n test -s amino_acids.phy</pre>
</div>
<p>Even though Alexandros Stamatakis is quite fond of the &#8220;CAT&#8221; models and they accelerate tree inference, they aren&#8217;t appropriate for use with pplacer.
We need to get an estimate of the gamma shape parameter.</p>
<p>PHYML can be run like so, on non-interleaved (hence the -q) phylip-format alignments:</p>
<div class="highlight-python"><pre>phyml -q -d nt -m GTR -i nucleotides.phy
phyml -q -d aa -m WAG -i amino_acids.phy</pre>
</div>
<p>Note that pplacer only works with phyml version 3.0 (the current version).</p>
<p>Both of these programs emit &#8220;statistics files&#8221;: files that describe the phylogenetic model used.
Pplacer then uses those same statistics to place your reads.
For RAxML, they are called something like <tt class="docutils literal"><span class="pre">RAxML_info.test</span></tt>, whereas for PHYML they are called something like <tt class="docutils literal"><span class="pre">test_aln_phyml_stats.txt</span></tt>.</p>
<p>If your taxon names have too many funny symbols, pplacer will get confused.
We have had a difficult time with the wacky names exported by the otherwise lovely software <a class="reference external" href="http://www.geneious.com/">geneious</a>.
If you have a tree which isn&#8217;t getting parsed properly by pplacer, and you think it should be, send it to us and we will have a look.</p>
<p>Avoid giving pplacer a reference tree with lots of very similar sequences.
It&#8217;s a waste of time&#8211; pplacer must evaluate the resultant branches like any others.
Identical sequences are especially bad, and the resultant zero length branches will make pplacer complain.</p>
<p>If you give pplacer a reference tree which has been rooted in the middle of an edge, you will get a warning like:</p>
<div class="highlight-python"><pre>Warning: pplacer results make the most sense when the given tree is multifurcating
at the root. See manual for details.</pre>
</div>
<p>In pplacer the two edges coming off of the root have the same status as the rest of the edges; therefore they will counted as two separate edges.
That will lead to artifactually low likelihood weight ratio and posterior probabilities for query sequences placed on those edges.
This doesn&#8217;t matter if your query sequences do not get placed next to the root, but you can avoid the problem altogether by rooting the tree at an internal node, or by leaving the outgroup in and rerooting the output trees.</p>
<div class="section" id="baseball">
<h3>Baseball<a class="headerlink" href="#baseball" title="Permalink to this headline">¶</a></h3>
<p>&#8220;Baseball&#8221; is one way that pplacer substantially increases the speed of placement, especially on very large trees.
Baseball is a game where the player with the bat has a certain number of unsuccessful attempts, called &#8220;strikes&#8221;, to hit the ball.</p>
<p>Pplacer applies this logic as follows.
Before placing placements, the algorithm gathers some extra information at each edge which makes it very fast to do a quick initial evaluation of those edges.
This initial evaluation of the edges gives the order with which those edges are evaluated in a more complete sense.
We will call full evaluations &#8220;pitches.&#8221;
We start with the edge that looks best from the initial evaluation; say that the ML attachment to that edge for a given query has log likelihood L.
Fix some positive number D, which we call the &#8220;strike box.&#8221;
We proceed down the list in order until we encounter the first placement which has log likelihood less than L - D, which we call a &#8220;strike.&#8221;
Continue, allowing some number of strikes, until we stop doing detailed evaluation of what are most likely rather poor parts of the tree.</p>
<p>You can control the behavior of baseball playing using the <tt class="docutils literal"><span class="pre">--max-strikes</span></tt>, <tt class="docutils literal"><span class="pre">--strike-box</span></tt>, and <tt class="docutils literal"><span class="pre">--max-pitches</span></tt> options.
If, for any reason, you wish to disable baseball playing, simply add <tt class="docutils literal"><span class="pre">--max-strikes</span></tt> to zero (this also disables the <tt class="docutils literal"><span class="pre">--max-pitches</span></tt> option).</p>
<p>Having control over these options raises the question of how to set them.
The answer to this question can be given by pplacer&#8217;s &#8220;fantasy baseball&#8221; feature.
To gain an understanding of the tradeoff between runtime and accuracy, it analyzes all <tt class="docutils literal"><span class="pre">--max-pitches</span></tt> best locations.
It then runs the baseball algorithm with each combination of strike box (from 0 to the specified <tt class="docutils literal"><span class="pre">--strike-box</span></tt>) and max strikes (from 1 to the specified <tt class="docutils literal"><span class="pre">--max-strikes</span></tt>).
Using these different settings the program reports</p>
<ul class="simple">
<li>the &#8220;batting average,&#8221; i.e. the number of times the baseball algorithm with those settings achieved the optimal location obtained by evaluating all <tt class="docutils literal"><span class="pre">--max-pitches</span></tt> best locations; found in the file prefix.batting_avg.out</li>
<li>the &#8220;log likelihood difference,&#8221; i.e. the difference between the ML log likelihood achieved by the baseball algorithm with those settings compared to the best obtained by evaluating all <tt class="docutils literal"><span class="pre">--max-pitches</span></tt> best locations; found in the file prefix.like_diff.out</li>
<li>the &#8220;number of trials,&#8221; i.e. the number of locations fully evaluated by the baseball algorithm with those settings; found in the file prefix.n_trials.out</li>
</ul>
<p>The fantasy mode is invoked by telling pplacer what average likelihood difference you would like via the <tt class="docutils literal"><span class="pre">--fantasy</span></tt> option.
You can also tell it to run an evenly-spaced fraction of the query sequences in fantasy mode using the <tt class="docutils literal"><span class="pre">--fantasy-frac</span></tt> option, which gives you an idea of the optimal run parameters for the rest of the sequences. For example:</p>
<div class="highlight-python"><pre>pplacer --max-strikes 10 --strike-box 10 --fantasy 0.05 --fantasy-frac 0.02 -r example.fasta...</pre>
</div>
<p>says to run pplacer trying all of the combinations of max strikes and strike box up to 10, looking for the optimal combination which will give an average log likelihood difference of 0.05, and running on 2% of the query sequences.
If, for any reason, you wish to disable baseball playing, simply add <tt class="docutils literal"><span class="pre">--max-strikes</span></tt> to zero (this also disables the <tt class="docutils literal"><span class="pre">--max-pitches</span></tt> option).</p>
<p>You can use R to plot these matrices in a heat-map like fashion like so:</p>
<div class="highlight-python"><pre>ba &gt; read.table("reads_nodups.batting_avg.out")
image(x=c(0:nrow(ba)-1),xlab= "strike box", ylab= "number of strikes", \
   y=c(1:ncol(ba)-1),z=as.matrix(ba), main="batting average")</pre>
</div>
<p>Note that we have set things up so that turning on posterior probability with <tt class="docutils literal"><span class="pre">-p</span></tt> now changes the default search parameters to make a deeper search as follows:</p>
<div class="highlight-python"><pre>--keep-at-most 20
--keep-factor 0.001
--max-strikes 20</pre>
</div>
<p>You can set these to anything you like by using these flags <em>after</em> the <tt class="docutils literal"><span class="pre">-p</span></tt>.</p>
</div>
<div class="section" id="fig-ranking">
<h3>Fig ranking<a class="headerlink" href="#fig-ranking" title="Permalink to this headline">¶</a></h3>
<p>&#8220;Fig ranking&#8221; is a way to reduce the number of initial comparisons done by
using the structure of the reference tree. This initial phase is not the
bottleneck for trees on a thousand or so taxa, but it is for trees on tens of
thousands of taxa or more.</p>
<p>If a value is specified as <tt class="docutils literal"><span class="pre">--fig-cutoff</span> <span class="pre">x</span></tt>, pplacer will find subtrees of
the reference tree (that we call figs) on the reference tree such that no two
leaves in the cluster have a distance of greater than <tt class="docutils literal"><span class="pre">x</span></tt>. Each leaf is
contained in exactly one fig. A representative edge of the fig is chosen as
follows: say <em>n</em> is the most proximal node contained in the fig; the
representative is the edge descending from <em>n</em> to the subtree with the greatest
number of leaves.</p>
<p>With a collection of figs, pplacer will rank each of the representative edges
by the initial evaluation likelihood given a query sequence. For each fig, in
this order, pplacer selects all of the edges within the fig as well as all of
the edges proximal to the fig up to the root of the tree. These edges are
ranked by the same initial evaluation likelihood before pplacer attempts to
place the query sequence on each in turn. No edge will be attempted twice; if
the same edge is proximal to two separate figs, it will only be attempted when
the first fig is evaluated.</p>
<p>As each fig is evaluated for a query sequence, pplacer will also select any
figs ordered immediately after the current fig where the difference between
that fig&#8217;s representative likelihood and the current fig&#8217;s representative
likelihood is less than the value of the <tt class="docutils literal"><span class="pre">--strike-box</span></tt> parameter. Each of
these figs&#8217; sets of edges are then merged into the current fig&#8217;s edge set.</p>
<p>To test the accuracy of fig evaluation vs. full evaluation, the
<tt class="docutils literal"><span class="pre">--fig-eval-all</span></tt> flag can be specified to do both fig and full evaluation,
then show the percentage of sequences where the best location chosen by full
evaluation and fig evaluation is the same. <tt class="docutils literal"><span class="pre">--fig-eval-all</span></tt> must be specified
to specify the <tt class="docutils literal"><span class="pre">--fig-eval-discrepancy-tree</span></tt> flag. If this flag is specified,
a tree will be written out showing the locations chosen by both methods for
each sequence where the two differ.</p>
<p>The colored trees written out by the <tt class="docutils literal"><span class="pre">--fig-tree</span></tt> and
<tt class="docutils literal"><span class="pre">--fig-eval-discrepancy-tree</span></tt> flags shows figs as colored subtrees within the
reference tree.</p>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="guppy.html" title="guppy"
             >next</a> |</li>
        <li class="right" >
          <a href="../index.html" title="pplacer documentation home"
             >previous</a> |</li>
        <li><a href="../index.html">pplacer 1.1 documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright 2011, Erick Matsen and Aaron Gallagher.
      Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 1.1.
    </div>
  </body>
</html>